src/mds/MDCache.cc

```diff
@@ -829,7 +831,92 @@ MDSCacheObject *MDCache::get_object(const MDSCacheObjectInfo &info)
 }
 
 
+// ====================================================================
+// consistent hash ring
+
+void MDCache::init_consistent_hash_ring()
+{
+  int rank = 0;
+  int max_mds = mds->mdsmap->get_max_mds();
+  int count = 0;
+  while (count < max_mds) {
+    add_rank_node_to_consistent_hash_ring(rank);
+    count++;
+  }
+}
+
+void MDCache::add_rank_node_to_consistent_hash_ring(mds_rank_t rank)
+{
+  uint32_t hash = rjhash32(rank);
```

Comment:
Why do ranks need to be hashed?

Logically the ranks are uniform "buckets" along the line (break the circle and stretch it out). Where each rank's bucket falls on that line is static based on the rank number and max_mds.

Since you're using `uint32_t` for the hash. If I had two ranks, rank 0 would get all inodes that hash between `[0, 2^31-1]` and rank 1 would get `[2^31, 2^32-1]`.

Reply:
For the way I've implemented it:
The ranks in the consistent hash ring are arranged in the ascending order of the hash(rank). Technically this hash need not be saved as part of the struct (because it can be calculated anytime) but when an inode is hashed and the rank needs to be found out (for migration), the algorithm works by doing a binary search on the rank hashes to find the lowest rank hash that's greater than or equal to the inode hash (inspired by the consistent hashing defined in Akamai's paper). Since this involves constant comparison of the rank hashes, I thought of storing the hash in memory instead of recalculating it every time.

I think your suggestion is another way of achieving the same concept of partitioning the hash space by having a range assigned to each rank.

Note: I think I should do a modulo(max no of mdss that cephfs can support) to reduce the hash space more.

Reply:
I believe the calculation is literally as simple as:

````
mds_rank_t MDCache::hash_into_rank_bucket(uint32_t hash, mds_rank_t max_mds)
{
    return std::min(hash / (UINT32_MAX / (uint32_t)max_mds), max_mds);
}
````

> The ranks in the consistent hash ring are arranged in the ascending order of the hash(rank).

I haven't looked but I think the original paper is more complicated because it's not dealing with buckets that are consecutive integers without holes. We can safely use the above method instead.

Reply:
Ahh! Got it now! The idea is the same as this paper https://arxiv.org/pdf/1406.2294.pdf I think.

Reply:
@batrick When max_mds changes, won't all the keys in the hash table get redistributed in the hashing scheme you suggested? I think we should follow something relating to this: https://arxiv.org/pdf/1406.2294.pdf (Jump Consistent Hash).